### PR TITLE
lang: allow the cfg attribute above the instructions

### LINKED ISSUE #2257: Anchor `#[program]` macro appears to ignore cargo feature flags
I expect to be able to use Cargo feature flags when developing a rust application. They are useful when following continuous integration allowing partially developed features to be integrated within a team to improve a teams workflow. 

Take the following example:

```rust
use anchor_lang::prelude::*;

declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");

#[program]
pub mod cfg_issue {
    use super::*;

    pub fn initialize(_ctx: Context<Initialize>) -> Result<()> {
        Ok(())
    }

    #[cfg(feature = "my-feature")]
    pub fn only_my_feature(_ctx: Context<Initialize>) -> Result<()> {
        Ok(())
    }
}

#[derive(Accounts)]
pub struct Initialize {}
```

With a configured `my-feature` feature flag in `./programs/cfg_issue/Cargo.toml`

```toml
# ...

[features]
my-feature = []

# ...

```

## Expected Behaviour

The program should compile as if the `only_my_feature` function was not in the source code.

## Actual Behaviour

The following will fail:

```bash
$ anchor build 
BPF SDK: /home/vscode/.local/share/solana/install/releases/1.10.39/solana-release/bin/sdk/bpf
cargo-build-bpf child: rustup toolchain list -v
cargo-build-bpf child: cargo +bpf build --target bpfel-unknown-unknown --release
   Compiling cfg-issue v0.1.0 (/workspaces/test-bug-anchor/cfg-issue/programs/cfg-issue)
error[E0425]: cannot find function `only_my_feature` in module `cfg_issue`
  --> programs/cfg-issue/src/lib.rs:14:12
   |
14 |     pub fn only_my_feature(_ctx: Context<Initialize>) -> Result<()> {
   |            ^^^^^^^^^^^^^^^ not found in `cfg_issue`
   |
   = note: consider importing this function:
           crate::__private::__global::only_my_feature

For more information about this error, try `rustc --explain E0425`.
error: could not compile `cfg-issue` due to previous error
```

Whilst including the feature makes the macro succeed as expected:

```bash
$ anchor build -- --features my-feature
BPF SDK: /home/vscode/.local/share/solana/install/releases/1.10.39/solana-release/bin/sdk/bpf
Features: my-feature
cargo-build-bpf child: rustup toolchain list -v
cargo-build-bpf child: cargo +bpf build --target bpfel-unknown-unknown --release --features my-feature
   Compiling cfg-issue v0.1.0 (/workspaces/test-bug-anchor/cfg-issue/programs/cfg-issue)
    Finished release [optimized] target(s) in 0.64s
cargo-build-bpf child: /home/vscode/.local/share/solana/install/releases/1.10.39/solana-release/bin/sdk/bpf/scripts/strip.sh /workspaces/test-bug-anchor/cfg-issue/target/bpfel-unknown-unknown/release/cfg_issue.so /workspaces/test-bug-anchor/cfg-issue/target/deploy/cfg_issue.so
cargo-build-bpf child: /home/vscode/.local/share/solana/install/releases/1.10.39/solana-release/bin/sdk/bpf/dependencies/bpf-tools/llvm/bin/llvm-readelf --dyn-symbols /workspaces/test-bug-anchor/cfg-issue/target/deploy/cfg_issue.so

To deploy this program:
  $ solana program deploy /workspaces/test-bug-anchor/cfg-issue/target/deploy/cfg_issue.so
The program address will default to this keypair (override with --program-id):
  /workspaces/test-bug-anchor/cfg-issue/target/deploy/cfg_issue-keypair.json
```

### PR DESCRIPTION
**Pull request by Aursen**
Closes #2257

The PR #2011 will solve the issue about the idl

### CODE REVIEW

**Comment by Henry-E** [comment]
Assuming this PR is essentially blocked until https://github.com/coral-xyz/anchor/pull/2011 is merged

File: lang/syn/src/codegen/program/accounts.rs [hunk 1]
```diff
@@ -3,7 +3,7 @@ use heck::SnakeCase;
 use quote::quote;
 
 pub fn generate(program: &Program) -> proc_macro2::TokenStream {
-    let mut accounts = std::collections::HashSet::new();
+    let mut accounts = std::collections::BTreeMap::new();
```
Comment (mina86) [inline lang/syn/src/codegen/program/accounts.rs:6]: Why not HashMap?

**Comment by mina86** [comment]
@Henry-E, now that #2011 is merged, can we have this merged?

File: lang/syn/src/codegen/program/accounts.rs (hunk 1 above)
Comment (Aursen) [reply lang/syn/src/codegen/program/accounts.rs:6]: > Why not HashMap?

I don't remember I did the PR a long time ago

Comment (Aursen) [reply lang/syn/src/codegen/program/accounts.rs:6]: But yeah it's better you right, we want to use this to cache values

**Review by acheroncrypto — COMMENTED** [review]
This looks good to me but one issue is that this only works with `idl-build` feature. Still, we should be able to get this in before the next release.

File: lang/syn/src/codegen/program/accounts.rs (hunk 1 above)
Comment (acheroncrypto) [reply lang/syn/src/codegen/program/accounts.rs:6]: It doesn't matter here actually, this is just macro generation code so it doesn't run on-chain.

**Review by acheroncrypto — CHANGES_REQUESTED** [review]
Returning back to this after https://github.com/coral-xyz/anchor/issues/3247. I think this looks great, but there are conflicts — are you able to fix them?

**Comment by Aursen** [comment]
> Returning back to this after https://github.com/coral-xyz/anchor/issues/3247. I think this looks great, but there are conflicts — are you able to fix them?

Let me do it this weekend

**Review by acheroncrypto — APPROVED** [review]
Thank you and sorry for the ~2 year delay!

**Pull request merged**